Level 3

nikerfe opened the pull request.

nikerfe committed:
создал регистрацию и авторизацию

nikerfe committed:
рендер карточек-movies,savedmovies(пофиксить иконку сохранения в movies)

nikerfe committed:
допилил все фичи + рефакторинг кода

SmilingJey commented:
Здравствуйте

В чеклисте есть требование, согласно которому, если "Фронтенд часть приложения не выложена на сервер;", то работа отклоняется от проверки https://yadi.sk/i/zvnaSnCVjhWKPw

Пожалуйста выложите проект на сервер, убедитесь, что он там работает корректно и приведите ссылку в комментарии к пулреквесту или readme файле

Сейчас работа отклоняется от проверки

nikerfe committed:
фиксы

nikerfe committed:
Update README.md

ссылка на сайт http://theworldofcinema.nomoredomains.monster/
IP-сайта: 178.154.202.238

nikerfe committed:
Update README.md

nikerfe committed:
фиксы preloader

nikerfe committed:
Merge branch 'level-3' of https://github.com/nikerfe/movies-explorer-frontend into level-3

nikerfe committed:
фиксы currentuser

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--30.27--> Функциональность проекта из брифа реализована полностью:
	- Прелоадер крутится в ожидании выполнения запроса.
	- Все ссылки и кнопки в проекте работают.
	- Все формы валидируются и на стороне клиента.
	- На странице «Регистрация» клик по кнопке «Зарегистрироваться» при корректно введённых данных отправляет запрос на роут `/signup`. Если запрос прошёл успешно, то происходит редирект на страницу `/movies`.
	**Комментарий: после успешной регистрации должен автоматически происходить вход и редирект на `/movies`**
	- На странице «Авторизации» клик по кнопке «Войти» при корректно введённых данных отправляет запрос на роут `/signin`. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- На странице редактирования профиля клик по кнопке «Сохранить» при корректно введённых данных отправляет запрос на роут `/users/me`- Если запрос прошёл успешно, пользователь получает соответствующее сообщение.
	**Комментарий: главная страница так же должна быть доступна авторизированному пользователю, на главной странице так же нужно менять состояние шапки когда пользователь авторизирован**
	- При работе с полями ввода на страницах «Регистрация» и «Авторизация» кнопки сабмита формы заблокированы и ожидают введения валидных данных в инпуты.
	- Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают, — появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт».
	- Работа с фильтром настроена. Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата. Запрос к API не выполняется повторно. Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	- После успешного сабмита формы поиска появляется блок с результатами. Если ничего не найдено, выводится надпись «Ничего не найдено».
	- В блоке результата отображаются 4 карточки. Нажатие на кнопку «Ещё» отображает следующие 4 карточки.
	**Комментарий: по заданию количество карточек должно зависеть от размера экрана, сейчас добавляются всегда 3 картчоки, а должен добавляться ряд карточек**
	- Карточка состоит из изображения, названия фильма и длительности. Длительность фильма рассчитывается корректно и соответствует формату, указанному в макете. Клик по карточке ведёт на трейлер фильма.
	- Если пользователь закрыл вкладку, а после — вернулся на сайт, данные достаются из локального хранилища при монтировании компонента `App`.
	**Комментарий: нужно подгружать и отображать ранее найденные и отфильтрованные фильмы на странице, по заданию "Как только запрос сделан, данные передаются в стейт-переменную и обновляются в локальном хранилище, а блок появляется. Если пользователь закрыл вкладку, а после — вернулся на сайт, нужно достать данные из локального хранилища при монтировании компонента App"**
	- При клике на иконку «Лайк» в блоке карточки выполняется запрос к `/movies` нашего API.
	- При клике на активную иконку «Лайк» в блоке карточки выполняется запрос на удаление карточки фильма. После успешного запроса карточка удаляется из раздела «Сохранённые фильмы».
	- На странице «Сохранённые фильмы» отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
	- На странице «Сохранённые фильмы» блок карточки содержит иконку удаления, а не лайка.
	- При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
- Отсутствуют серьёзные баги интерфейса:
	- [x] <!--1.52--> Удалённые/добавленные карточки фильмов появляются только после перезагрузки страницы. 
	- [x] <!--1.52--> Фильтрация по короткометражкам срабатывает после перезагрузки страницы. 
	- [x] <!--1.52--> Иконка «Лайк» не меняет свой статус после перезагрузки страницы. 
	- [x] <!--1.52--> При пустых полях ввода кнопка сабмита формы активна. 
	- [x] <!--1.52--> Когда отображены все фильмы по ключевому слову, кнопка «Ещё» скрывается только после перезагрузки страницы. 
	- [x] <!--1.52--> Если осталось отобразить менее 4 карточек фильмов, то нажатие кнопки «Ещё» вызывает появления ошибок в консоли. 
	- [x] <!--1.52--> При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы. 
	- [x] <!--1.52--> Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или приложения. 
- [x] <!--1.52--> Сборка и запуск проекта выполняются без ошибок.  
- [x] <!--1.52--> Создана глобальная стейт-переменная `currentUser`.
- [x] <!--1.52--> Локальное хранилище: работа с `jwt` токеном и результатами поискового запроса выполнена в правильном порядке. 
- [x] <!--1.52--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище
- Регистрация и авторизация:
	- [x] <!--1.52--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. 
	- [x] <!--1.52--> При попытке перейти на любой защищённый роут происходит редирект на `/` . 
	- [ ] <!--1.52--> Если пользователь закрыл вкладку и был авторизован, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации. 
	**Комментарий: при повторном переходе на любую страницу происходит перенаправление на /movies**	
	- [x] <!--1.52--> После успешной авторизации происходит редирект на `/movies`. Шапка сайта отвечает состоянию авторизованного пользователя и содержит все необходимые ссылки. 
	- [x] <!--1.52--> Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована. 
	- [x] <!--1.52--> Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
	- [x] <!--1.52--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`. 
	- [x] <!--1.52--> При выходе из аккаунта `jwt` удаляется. 
	- [x] <!--1.52--> Корректно используется хук `useHistory`. 
	- [x] <!--1.52--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Компоненты:
	- [x] <!--1.52--> Хуки не используются внутри условных блоков или циклов. 
	- [x] <!--1.52--> Хуки вызываются в основной функции компонента. 
	- [x] <!--1.52--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
	- [x] <!--1.52--> Для элементов списка используется уникальный ключ `key`, а не индекс массива;
	- [x] <!--1.52--> Компоненты `Main`, `NavBar` и `SavedMoviesHeader` подписаны на контекст `CurrentUserContext`. 
	- [x] <!--1.52--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
	- [x] <!--1.52--> В корневом компоненте `App` создана стейт-переменная `currentUser`. Она используется в качестве значения для провайдера контекста. 
	- [x] <!--1.52--> В корневом компоненте `App` описаны обработчики: `onRegister`, `onLogin`, `onSignOut` и `onUpdateUser`. Эти обработчики вынесены в `App` и переданы в соответствующие компоненты: `Register.js`, `Login.js` и `NavBar.js`. 
- [x] <!--1.52--> Асинхронные запросы к API:
    - Запросы можно осуществлять через `Fetch API` или `XMLHttpRequest`.
	- Сторонние библиотеки (такие как `axios` или `jQuery`) не применяются.
	- Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`. 
	- Цепочка обработки промисов завершается блоком catch.
	- Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
- [x] <!--1.52--> Именование:
	- `camelCase` для параметров, методов и переменных;  
	- существительные для имён переменных;  
	- существительные во множественном числе для `NodeList`;  
	- существительные с заглавной буквы для имён классов; 
	- `const` для неизменяемых напрямую переменных; 
	- имя класса соответствует содержимому;
	- для именования запрещены: транслит и неуместные сокращения.
- [x] <!--1.52--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.

### Хорошие практики
- [x] <!--1.88--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--1.88--> Запрос к API за информацией о пользователе и массиве карточек выполняется единожды при монтировании. 
- [x] <!--1.88--> Запросы к API описаны внутри компонента App или в корневых компонентах страниц. 
- [x] <!--1.88--> В качестве функции-обработчика не передаются стрелочные функции. 
- [x] <!--1.88--> Обработка ошибок API:
  - [x] <!--1.88--> пользователь получает сообщение в случае любой ошибки; 
  - [ ] <!--1.88--> поля формы заблокированы во время отправки запросов. 
  **Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [ ] <!--1.88--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще" на разных размерах экрана и точки перестроения экрана нужно вынести в константы**

### Рекомендации
- [x] <!--1.67--> Используются семантически правильные блоки для компонентов.
- [x] <!--1.67--> Не используются <div> или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--1.67--> Отсутствует «мусор» в коде :
  - по имени коллбэка слушателя понятно его назначение; 
  - нет «магических значений» — все значения записаны в переменные;
  - константы описаны в той области видимости, в которой объявлена функция. Внутри функции значения констант берутся из замыкания. 

## Доп. комментарий:
- при дизлайке некоторых фильмов падает 403 ошибка https://yadi.sk/i/h6OZGmjpnfjqZQ , пользователь должен видеть только сохраненные им фильмы, поэтому 403 ошибка, которая появляется при удалении некоторых фильмов не должна возникать т.к. фильмов другого пользователя просто не должно быть, проверьте работу бэкенда 

## Количество баллов: 65

## Хорошая работа, большая часть задания работает верно, но баллов пока недостаточно для сдачи, нужно поправить как минимум замечания из раздела "Функциональность проекта из брифа реализована полностью"

nikerfe committed:
правки после ревью

nikerfe committed:
правки после ревью

nikerfe committed:
правки после ревью

nikerfe commented:
Поправил замечания с комментариями

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--30.27--> Функциональность проекта из брифа реализована полностью:
	- Прелоадер крутится в ожидании выполнения запроса.
	- Все ссылки и кнопки в проекте работают.
	- Все формы валидируются и на стороне клиента.
	- На странице «Регистрация» клик по кнопке «Зарегистрироваться» при корректно введённых данных отправляет запрос на роут `/signup`. Если запрос прошёл успешно, то происходит редирект на страницу `/movies`.
	- На странице «Авторизации» клик по кнопке «Войти» при корректно введённых данных отправляет запрос на роут `/signin`. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- На странице редактирования профиля клик по кнопке «Сохранить» при корректно введённых данных отправляет запрос на роут `/users/me`- Если запрос прошёл успешно, пользователь получает соответствующее сообщение.
	- При работе с полями ввода на страницах «Регистрация» и «Авторизация» кнопки сабмита формы заблокированы и ожидают введения валидных данных в инпуты.
	- Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают, — появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт».
	- Работа с фильтром настроена. Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата. Запрос к API не выполняется повторно. Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	- После успешного сабмита формы поиска появляется блок с результатами. Если ничего не найдено, выводится надпись «Ничего не найдено».
	- В блоке результата отображаются 4 карточки. Нажатие на кнопку «Ещё» отображает следующие 4 карточки.
	- Карточка состоит из изображения, названия фильма и длительности. Длительность фильма рассчитывается корректно и соответствует формату, указанному в макете. Клик по карточке ведёт на трейлер фильма.
	- Если пользователь закрыл вкладку, а после — вернулся на сайт, данные достаются из локального хранилища при монтировании компонента `App`.
	- При клике на иконку «Лайк» в блоке карточки выполняется запрос к `/movies` нашего API.
	- При клике на активную иконку «Лайк» в блоке карточки выполняется запрос на удаление карточки фильма. После успешного запроса карточка удаляется из раздела «Сохранённые фильмы».
	- На странице «Сохранённые фильмы» отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
	- На странице «Сохранённые фильмы» блок карточки содержит иконку удаления, а не лайка.
	- При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
- Отсутствуют серьёзные баги интерфейса:
	- [x] <!--1.52--> Удалённые/добавленные карточки фильмов появляются только после перезагрузки страницы. 
	- [x] <!--1.52--> Фильтрация по короткометражкам срабатывает после перезагрузки страницы. 
	- [x] <!--1.52--> Иконка «Лайк» не меняет свой статус после перезагрузки страницы. 
	- [x] <!--1.52--> При пустых полях ввода кнопка сабмита формы активна. 
	- [x] <!--1.52--> Когда отображены все фильмы по ключевому слову, кнопка «Ещё» скрывается только после перезагрузки страницы. 
	- [x] <!--1.52--> Если осталось отобразить менее 4 карточек фильмов, то нажатие кнопки «Ещё» вызывает появления ошибок в консоли. 
	- [x] <!--1.52--> При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы. 
	- [x] <!--1.52--> Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или приложения. 
- [x] <!--1.52--> Сборка и запуск проекта выполняются без ошибок.  
- [x] <!--1.52--> Создана глобальная стейт-переменная `currentUser`.
- [x] <!--1.52--> Локальное хранилище: работа с `jwt` токеном и результатами поискового запроса выполнена в правильном порядке. 
- [ ] <!--1.52--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище
**Комментарий: при неверном токене не происходит редирект на вход https://yadi.sk/i/aunFbQIcdfmAqg**
- Регистрация и авторизация:
	- [x] <!--1.52--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. 
	- [x] <!--1.52--> При попытке перейти на любой защищённый роут происходит редирект на `/` . 
	- [x] <!--1.52--> Если пользователь закрыл вкладку и был авторизован, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации. 
	- [x] <!--1.52--> После успешной авторизации происходит редирект на `/movies`. Шапка сайта отвечает состоянию авторизованного пользователя и содержит все необходимые ссылки. 
	- [x] <!--1.52--> Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована. 
	- [x] <!--1.52--> Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
	- [x] <!--1.52--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`. 
	- [x] <!--1.52--> При выходе из аккаунта `jwt` удаляется. 
	- [x] <!--1.52--> Корректно используется хук `useHistory`. 
	- [x] <!--1.52--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Компоненты:
	- [x] <!--1.52--> Хуки не используются внутри условных блоков или циклов. 
	- [x] <!--1.52--> Хуки вызываются в основной функции компонента. 
	- [x] <!--1.52--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
	- [x] <!--1.52--> Для элементов списка используется уникальный ключ `key`, а не индекс массива;
	- [x] <!--1.52--> Компоненты `Main`, `NavBar` и `SavedMoviesHeader` подписаны на контекст `CurrentUserContext`. 
	- [x] <!--1.52--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
	- [x] <!--1.52--> В корневом компоненте `App` создана стейт-переменная `currentUser`. Она используется в качестве значения для провайдера контекста. 
	- [x] <!--1.52--> В корневом компоненте `App` описаны обработчики: `onRegister`, `onLogin`, `onSignOut` и `onUpdateUser`. Эти обработчики вынесены в `App` и переданы в соответствующие компоненты: `Register.js`, `Login.js` и `NavBar.js`. 
- [x] <!--1.52--> Асинхронные запросы к API:
    - Запросы можно осуществлять через `Fetch API` или `XMLHttpRequest`.
	- Сторонние библиотеки (такие как `axios` или `jQuery`) не применяются.
	- Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`. 
	- Цепочка обработки промисов завершается блоком catch.
	- Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
- [x] <!--1.52--> Именование:
	- `camelCase` для параметров, методов и переменных;  
	- существительные для имён переменных;  
	- существительные во множественном числе для `NodeList`;  
	- существительные с заглавной буквы для имён классов; 
	- `const` для неизменяемых напрямую переменных; 
	- имя класса соответствует содержимому;
	- для именования запрещены: транслит и неуместные сокращения.
- [x] <!--1.52--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.

### Хорошие практики
- [x] <!--1.88--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--1.88--> Запрос к API за информацией о пользователе и массиве карточек выполняется единожды при монтировании. 
- [x] <!--1.88--> Запросы к API описаны внутри компонента App или в корневых компонентах страниц. 
- [x] <!--1.88--> В качестве функции-обработчика не передаются стрелочные функции. 
- [x] <!--1.88--> Обработка ошибок API:
  - [x] <!--1.88--> пользователь получает сообщение в случае любой ошибки; 
  - [ ] <!--1.88--> поля формы заблокированы во время отправки запросов. 
  **Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, что бы пользователь не мог выполнить новые запросы до завершения предыдущего**
- [x] <!--1.88--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--1.67--> Используются семантически правильные блоки для компонентов.
- [x] <!--1.67--> Не используются <div> или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--1.67--> Отсутствует «мусор» в коде :
  - по имени коллбэка слушателя понятно его назначение; 
  - нет «магических значений» — все значения записаны в переменные;
  - константы описаны в той области видимости, в которой объявлена функция. Внутри функции значения констант берутся из замыкания. 

## Доп. комментарий:
- при дизлайке некоторых фильмов падает 403 ошибка https://yadi.sk/i/h6OZGmjpnfjqZQ , пользователь должен видеть только сохраненные им фильмы, поэтому 403 ошибка, которая появляется при удалении некоторых фильмов не должна возникать т.к. фильмов другого пользователя просто не должно быть, проверьте работу бэкенда 

## Количество баллов: 97

## Отлично, теперь необходимый функционал работает. Поздравляю с завершением дипломного проекта!

### Несколько советов для дальнейшего развития:
- свертайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- еще немного попрактивокаться с React, например изучить redux
- будет большим плюсом при поиске работы, если немного изучите TypeScript т.к. на него переходят многие компании
- ходить на собеседования, опыт прохождения собеседований не менее важен знаний теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушаете подкаксты, они очень хорошо расширяют кругозор, и позволяют узнать, что проиходит в отрасли:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw
